http-gateways/PATH_GATEWAY.md

```diff
@@ -701,5 +701,9 @@ The usual optimizations involve:
 
 ## Streaming errors
 
-Gateways MUST force-close HTTP connections if they detect an error while
-streaming a file to avoid that users receive incomplete, yet valid, files.
+To avoid users receiving an incomplete, yet valid, files, the gateway MUST
+close the HTTP stream if an error occurs while streaming a file to the client.
+This can be done via the following mechanisms:
+
+- Sending a `RST` (reset) frame for HTTP/1.1
+- Sending a `RST_STREAM` (reset stream) frame for HTTP/2
```

Comment:
```suggestion
- Sending a `RST_STREAM` (reset stream) frame for HTTP/2
- Sending a `CANCEL_PUSH` frame for HTTP/3, see RFC9114 section A.2.5.3.
```

Reply:
I haven't read the entire spec, but according to [BCP 56](https://www.rfc-editor.org/info/bcp56), HTTP APIs shouldn't go into this level of detail. HTTP is designed in a way that allows building applications without referring to the specific HTTP version.

Reply:
I didn't actually red the RFC, I know RST_STREAM on HTTP/2 is what we want, I just CTRL + F RST_STREAM in the HTTP/3 RFC and found this text:
>    RST_STREAM (0x03):  RST_STREAM frames do not exist in HTTP/3, since
      QUIC provides stream lifecycle management.  The same code point is
      used for the CANCEL_PUSH frame (Section 7.2.3).

> HTTP is designed in a way that allows building applications without referring to the specific HTTP version.

Streaming errors is not a thing HTTP supports, you can setup trailler headers but they aren't supported by browsers.
So we are stealing the rug from under HTTP and use the underlying protocol to generate an unexpected error (which is actually supported and will return errors in all client implementations I've tested), that why we need to use TCP FIN.

I don't know if we should add this to the gateway spec, because I don't think most server implementations give you as much control as the go std does.
How will you implement that on reverse proxies ?

Reply:
@marten-seemann as @Jorropo mentioned, the big issue here is that HTTP doesn't have any error handling for streaming. One of my motivations behind this IPIP regards the new TAR format for the gateway (https://github.com/ipfs/kubo/pull/9029).

It is possible that the TAR creation fails while streaming the file to the client due to many reasons. However, if you just stop streaming the TAR, you still get a valid TAR file, but it is incomplete. There's no feedback. Printing a trailer header is useless since browsers are not able to parse them. The only way we found so far to tell the user that something **is** wrong was by force-closing the HTTP stream.

I also have mixed feelings about having this on the spec since it is so specific. In addition, as @Jorropo mentioned it may be worth it having an opt-out of the behaviour through some header.

Reply:
@hacdias 
> How will you implement that on reverse proxies ?

I think this is an important question.
I would like to see something like: `X-On-Error: reset` header sent by the server to indicate this will be done, then a reverse proxy could just not send thoses headers, and clients would be able to tell it is not gonna do this.

Reply:
@Jorropo please see the updates I've made to the IPIP.